public/app/features/explore/state/utils.ts

```diff
@@ -119,12 +123,64 @@ export async function loadAndInitDatasource(
     }
   }
 
-  const historyKey = `grafana.explore.history.${instance.meta?.id}`;
-  const history = store.getObject<HistoryItem[]>(historyKey, []);
-  // Save last-used datasource
+  let history: HistoryItem[] = [];
+
+  const localStorageHistory = getLocalRichHistoryStorage();
+
+  const historyResults = await localStorageHistory.getRichHistory({
+    search: '',
+    sortOrder: SortOrder.Ascending,
+    datasourceFilters: [instance.name],
```

Comment:
When mixed is selected as the root, could we filter by multiple data sources in here? 🤔  At the moment I don't see history for a data source that is used as a child of mixed (in the video: when switched to mixed I still could see "created in non-mixed mode" entry):


https://github.com/grafana/grafana/assets/745532/2c19f8b8-6193-4e42-8549-21c857a58c6c

Reply:
@ifrost Its not impossible but I'd like to make that a separate piece of work due to its complexity and the necessity of changing a core component.

The logic to bring history into queries is [loadAndInitDatasource](https://github.com/grafana/grafana/blob/main/public/app/features/explore/state/utils.ts#L99) which only brings in a single datasource, and returns only that datasource's history. We could have places that call this function also pass in the datasources available in the queries, and then return those histories as well. What do you think about a ticket doing the following:

Change loadAndInitDatasource to have as parameters
- orgId: number,
- rootDatasource: DataSourceRef | string
- queryDatasources? DataSourceRef | string[]

and return `Promise<{ queryHistories: Array<{ refId: string; history: HistoryItem[]}>; rootInstance: DataSourceApi }>`

With this change, if (for example) the root datasource is mixed with loki as the datasource for query refId A, it will be called with `(1, mixed, [loki])` and return `[{refId: 'A', history: [loki history]}]` as the history. 

We will need to modify how history is kept in state, as it is now just assuming one history for query https://github.com/grafana/grafana/blob/main/public/app/types/explore.ts#L126

I think the structure `Array<{ refId: string; history: HistoryItem[]}>` would work fine for state as well.

Finally, explore now retrieves one history for every row ([src](https://github.com/grafana/grafana/blob/main/public/app/features/explore/QueryRows.tsx#L44)) and passes it onto the core component `QueryEditorRows`. `QueryEditorRows` attaches the same history to every query ([src](https://github.com/grafana/grafana/blob/main/public/app/features/query/components/QueryEditorRows.tsx#L179)). But we could make it attach the right history for the given RefId - as long as we're sure to handle the history change if the query datasource changes.

Let me know if there's something I'm missing, if you'd like me to tackle this in this PR, or if I should write up an issue. I can get started on that right away.

Reply:
Oh, a drawback to this is that there will be some redundancy - like if RefId A and B use the same loki datasource, it would have the history for both seperately. It is more straightforward than querying off datasource UID but that would be more efficient. Let me know if youd rather I do that.

Reply:
Oh I thought it was as simple as changing `datasourceFilters: [instance.name],` to use names from each query but if I understand correctly we would need to load the data source to get it's name. 

The current implementation makes things better anyway, so I'm happy to keep  not include it in this PR.